fixes #3912 - add inheritance for locations / organizations

isratrade opened the pull request.

isratrade commented on app/assets/stylesheets/application.scss:
```diff
@@ -203,6 +203,7 @@ select{padding: initial;}
 .collapse .inputs-list li{ margin-left:20px; }
 
 .delete{color: #b94a48 !important; cursor: auto;}
+.inherited{font-weight: bold !important; cursor: auto;}
```
this is just to differentiate disabled items that are disabled because used by host or disabled because inherited.  There is is a tooltip if disabled because inherited

isratrade commented on app/controllers/concerns/foreman/controller/taxonomies_controller.rb:
```diff
@@ -70,8 +70,10 @@ def step2
     result = Taxonomy.no_taxonomy_scope do
-      (params[taxonomy_single.to_sym][:ignore_types] -= ["0"]) if params[taxonomy_single.to_sym][:ignore_types]
-      @taxonomy.update_attributes(params[taxonomy_single.to_sym])
+      (params[taxonomy_single]["ignore_types"] -= ["0"]) if params[taxonomy_single]["ignore_types"]
```
I'm using string keys in all places. I had occurrences of mismatches of comparing string and symbol keys

ohadlevy replied:
i think this is due another issue, maybe not relevant after #1125

isratrade commented on app/views/taxonomies/_form.html.erb:
```diff
@@ -23,86 +23,98 @@
-        <%= multiple_selects f, :users, User.except_admin, taxonomy.selected_ids[:user_ids],
-           {:disabled => taxonomy.used_and_selected_ids[:user_ids], :label => _("Users")},
-           {'data-mismatches' => taxonomy.need_to_be_selected_ids[:user_ids].to_json} %>
+        <%= multiple_selects f, :users, User.except_admin, taxonomy.selected_or_inherited_ids["user_ids"],
```
change from `selected_ids` to `selected_or_inherited_ids`
change from `used_and_selected_ids` to `used_and_selected_or_inherited_ids`

isratrade commented on vendor/assets/javascripts/jquery.multi-select.js:
```diff
@@ -502,5 +508,13 @@ function multiSelectOnLoad(){
+    if (!(inheriteds == null || inheriteds == 'undefined')) {
+      var inherited_ids = $.parseJSON(inheriteds);
+      $.each(inherited_ids, function(index,inherited_id){
+        opt_id = (inherited_id +"").replace(/[^A-Za-z0-9]*/gi, '_')+'-selection';
```
inheriteds uses `-selection` here where mistmaches uses `-selectable`

isratrade commented on vendor/assets/javascripts/jquery.multi-select.js:
```diff
@@ -502,5 +508,13 @@ function multiSelectOnLoad(){
+      var inherited_ids = $.parseJSON(inheriteds);
+      $.each(inherited_ids, function(index,inherited_id){
+        opt_id = (inherited_id +"").replace(/[^A-Za-z0-9]*/gi, '_')+'-selection';
+        $('#ms-'+$(item).attr('id')).find('#'+opt_id).addClass('inherited').tooltip({title: _("This is inherited from parent"), placement: "right"});
```
add css class 'inherited' that makes it bold with a tooltip.

isratrade commented on app/models/concerns/taxonomix.rb:
```diff
@@ -23,7 +23,7 @@ def with_taxonomy_scope
 
     def inner_select taxonomy
-      taxonomy_ids = Array.wrap(taxonomy).map(&:id).join(',')
+      taxonomy_ids = taxonomy.path_ids.join(',')
```
@abenari, is this change ok?  It what cases previously was `taxonomy` more than one.  I remember I challenged this in the past, but there were cases where `taxonomy` was more than a single instance.  Otherwise, Array.wrap and .map is not needed.

ohadlevy replied:
@abenari ping ?

ohadlevy replied:
afair, there was one case where the form would submit multiple orgs/locations, however I cant remember where that was needed, @abenari does that ring a bell?

ares replied:
Maybe http://projects.theforeman.org/issues/2298 and https://github.com/theforeman/foreman/pull/864 will help to make it clearer. I think user can have more orgs/locs when he logs in and does not select any (which is considered a bug).

isratrade commented:
re-committed. added some tests and ready for review

ohadlevy commented on app/controllers/concerns/foreman/controller/taxonomies_controller.rb:
```diff
@@ -35,6 +35,14 @@ def new
 
+  def nest
+    @taxonomy = taxonomy_class.new
+    Taxonomy.no_taxonomy_scope do
```
why without scope? shouldnt it be the parent scope or something like that? could this be abused?

ohadlevy commented on app/helpers/hosts_and_hostgroups_helper.rb:
```diff
@@ -1,14 +1,15 @@
-    nesting = hostgroup.label.to_s.gsub(/[^\/]+\/?$/, "")
-    nesting = truncate(nesting, :length => max_length - hostgroup.name.to_s.size) if nesting.to_s.size > 0
-    name =  truncate(hostgroup.name, :length => max_length - nesting.to_s.size)
+  # this helper is used for hostgroup names and location/organization names.  Both have ancestry
```
rename the file then?

isratrade replied:
@ohadlevy, I'm not sure it's necessary to rename the file. Most methods relate to host/hostgroup. All helper methods are available to all views.

ohadlevy commented on app/models/taxonomy.rb:
```diff
@@ -1,12 +1,15 @@
 class Taxonomy < ActiveRecord::Base
   audited :allow_mass_assignment => true
   has_associated_audits
+  has_ancestry :orphan_strategy => :restrict
```
this can generate exceptions, see #924

ohadlevy commented on app/models/taxonomy.rb:
```diff
@@ -125,10 +130,12 @@ def dup
-  private
+  def get_label
+    return name if ancestry.empty?
+    ancestors.map{|a| a.name + "/"}.join + name
```
is this the same as in hostgroups? maybe extract to a mixin/class?

ohadlevy commented on app/services/tax_host.rb:
```diff
@@ -22,7 +22,7 @@ def used_ids
     ids.default = []
     hash_keys.each do |col|
-      ids[col] = self.send(col)
+      ids[col.to_s] = self.send(col)
```
why not use HashWithIndifferentAccess instead ?

isratrade commented:
@ohadlevy, do you want this feature merged soon?

ohadlevy commented:
@isratrade yes please :)

isratrade commented:
@abenari, I have one failure caused by `test/unit/classification_test.rb:38` 
It seems to have to do with the matcher but I didn't add inheritance in the fixtures, so I'm confused what changed.

```
<"test"> expected but was
<"secret">. (test_0004_enc_should_return_updated_cluster_param)
/var/lib/workspace/workspace/test_develop_pull_request@4/database/sqlite3/puppet/3.0/ruby/1.8.7/test/unit/classification_test.rb:38    
```

ohadlevy commented:
@isratrade ping?

isratrade commented:
@ohadlevy, I don't understand the test failure to classification_test.  It would be good if another set of eyes could look at it.

isratrade commented:
@ohadlevy, classification_test now passes after some not changing `to_label`.  Note that without PR #1104, the inherited items from parent (on the right selected side) can be unselected which is not good.

isratrade commented:
re-committed and access_permission_test now passes. [test]

isratrade commented:
same issue.  failure only on mysql / ruby 2 that I don't think is relevant

isratrade commented:
Added new commit to disable used locations/organizations on edit form for domain, subnet, etc. I also included the js needed in PR #1104

isratrade commented:
re-committed with fixes to taxonomix.rb that caused failing tests

domcleal commented:
@isratrade ignore the sqlite/1.9.3 error, but you have integration test failures

isratrade commented:
@domcleal, not sure about the integration failures that popped up.  I can't reproduce them locally.  I'm continue tomorrow.

isratrade commented on app/controllers/application_controller.rb:
```diff
@@ -17,6 +17,7 @@ class ApplicationController < ActionController::Base
   before_filter :set_taxonomy, :require_mail, :check_empty_taxonomy
   before_filter :welcome, :only => :index, :unless => :api_request?
   before_filter :authorize
+  before_filter :clean_taxonomy_ids, :only => [:create, :update]
```
this is needed, since the the locations / organizations tab was switch to use multi-select.js and apparently this js library sends duplicate ids
ex.
params[:domain][:location_ids] = [","45", "45"] even though the [selected_ids] array does not have any duplicates.

ohadlevy commented on app/models/hostgroup.rb:
```diff
@@ -182,4 +160,10 @@ def remove_duplicated_nested_class
+  # overwrite method in taxonomix, since hostgroup has ancestry
+  def used_taxonomy_ids(type)
+    return [] if new_record? && parent_id.blank?
+    Host::Base.where(:hostgroup_id => self.path_ids).pluck(type).uniq
```
missing compact?

isratrade replied:
@ohadlevy, fixed and re-committed

isratrade commented:
there are still two test failures on cloning location / organization because Smart Proxy fixtures are not valid due to duplicate URL's

isratrade commented on app/models/host/managed.rb:
```diff
@@ -55,8 +55,8 @@ class Jail < ::Safemode::Jail
-      conditions[:organization_id] = Array.wrap(org).map(&:id) if org
-      conditions[:location_id]     = Array.wrap(loc).map(&:id) if loc
+      conditions[:organization_id] = org.path_ids if org
+      conditions[:location_id]     = loc.path_ids if loc
```
all tests should now pass.  I also made this change above and squashed it.

isratrade commented:
added new commit. changes are.
1) context of viewing of hosts (or other objects), scope by location / organization and their children (not parents)
2) context of creating new host, scope objects by location / organization and their parents
[test]

isratrade commented:
@domcleal, it looks like the tests passed, despite the 'red' for sqlite and 1.9.3

isratrade commented:
@domcleal, rebased PR.  branched develop and cherry-picked the commits.  I hope this helps jenkins

isratrade commented:
rebased

isratrade commented:
@domcleal, fyi, github bug, green on Conversation tab, but yellow circle on Commits tab

ares commented on app/views/users/_form.html.erb:
```diff
@@ -55,7 +55,7 @@
             <%= f.select :organizations_andor, [[_("must be"), "and"], [_("plus all"), "or"]], {}, :class => "form-control" %>
           </div>
-          <%= multiple_checkboxes f, :organizations, @user, Organization, :label=>'' %>
+          <%= location_selects f, @user.used_or_selected_organization_ids, {:disabled => @user.used_organization_ids, :label => ''} %>
```
organization_selects probably?

ares commented on app/controllers/hosts_controller.rb:
```diff
@@ -572,11 +572,11 @@ def taxonomy_scope
     if SETTINGS[:organizations_enabled]
       @organization ||= Organization.current
-      @organization ||= Organization.my_organizations.first
+      @organization ||= Organization.my_organizations.reorder(:label).first
```
Out of curiosity, why do we need to reorder here? Seems like my_\* alread orderd by label,name. Reorder is on many places in this PR.

ares commented on app/models/concerns/taxonomix.rb:
```diff
@@ -15,22 +15,46 @@ module Taxonomix
+      # In the case of users we want the taxonomy scope to get both the users of the taxonomy and admins.
+      inner_ids << admin_ids if inner_ids && self == User
+      scope = scope.where(:id => inner_ids) if inner_ids
       scope.readonly(false)
```
so we find all records within specific location, the find the same for specific organization and then combine these ids which we use in third query? sounds like work for SQL instead of ruby? I understand it's not that easy using ancestry but this does not feel right to me, what do others think?

isratrade replied:
@ares, I added issue http://projects.theforeman.org/issues/4329 to 
refactor `with_taxonomy_scope` to optimize SQL

ares commented on app/models/concerns/taxonomix.rb:
```diff
@@ -15,22 +15,46 @@ module Taxonomix
+    end
+
+    def inner_select(taxonomy, inner_method)
+      # always include ancestor_ids in inner select
```
Why do we want to include all ancestors? I tested on following example, I have A < B < C locations, current context is B, I have one subnet set to A, second to C. On subnets page I see both because we use :subtree_ids method, so all children + ascendants. Is this what we want? I thought we go only by one direction every time.

isratrade replied:
@ares, for the simple example of A > B > C, then yes we see both.  However, for multiple  branches, the ancestor_ids is a always straight line up, but the opposite direction, subtree_ids, is not always the same path down.
A
  --B1
    -- C1 - D1 - E1 
  --B2
    --  C2 - D2
  --B3
    --  C3 -  D3

If my current location is B1, I see the ancestors (A) and descendants (C1, D1, E1), but I don't see B2, B3, C2, C3, D2, D3.  However, If my current location was A, then I would see all.  Does this make sense?

ares replied:
@isratrade I understand how it works, but I don't understand why it works this way. A > B > C was simplest example where I wanted to show that we 'inherit' in both directions. So when I select B1 (Brno) I will see also all subnets assigned to A (Czech Republic) and C1 (Building 1) and C2 (Building 2). I'm confused, why should I see all of these? I think I shouldn't touch subnet from A (Czech Republic) when I'm in B1 (Brno) since I don't know what impact will it have on others B2 (e.g. Prague). Maybe I just don't see some usecases but 'inheriting' both directions seems strange to me.

isratrade replied:
@ares, I understand the confusion.  I initial thought similarly.  However, it goes back to the idea of white-listing vs black-listing or children add from their parents or delete from their parents. We chose that children inherit from the parents and can add to them.  Therefore, B1 (Brno) inherits from A (Czech Republic) and thus can see what's in A and add to it.  

The reason for both directions to do fix the awkward case (that I originally demoed), where Israel/Tel Aviv/Rack7 shows all domains but Israel only shows very few domains.  Intuitively, it should be opposite.  If Israel is current location, I should see all under Israel, wheras if I'm in "Rack7", I should see everything that Rack7 inherited PLUS its children (if any). Thus, both directions.

domcleal replied:
:+1: for @isratrade's argument, this makes sense.

domcleal replied:
I've been thinking about this some more, and it's going to get confusing when permissions are mapped onto inherited taxonomies.  This might be where @ares is coming from.

Consider that A (Czech Republic) has cz.redhat.com assigned to it, and we want everybody within CZ nested locations to be able to use it.  We also define brq.cz.redhat.com and assign it to B (Brno).

We then create @ares a user account and assign him edit permissions to Brno.  In my head, I intend for him to only be able to edit brq.cz.redhat.com, but to be able to view and use cz.redhat.com.

I don't think we want view/edit permissions to work differently, so the solution (subject to the new filters system) appears to me to give him two permissions:
- view_domains to cz.redhat.com
- edit_domains to brq.cz.redhat.com

And if you do that, you don't want to inherit objects from parents automatically, right?  You'd make it explicit as two separate filters, and only associate to the taxonomy you're assigned to and those below it.

isratrade replied:
Right, @ares agreed that permission filters don't inherit from parents automatically.  You need to make two separate filters.

ares replied:
Yes, that was our conclusion. Inheritance is not applied on filters. Filters are loaded from taxonomies that were used for resource searching (domain in this example) which uses inheritance. So we'd find both permissions and each would be applied to one domain. I'm happy you came to the same findings.

ares commented on app/models/taxable_taxonomy.rb:
```diff
@@ -11,4 +12,16 @@ class TaxableTaxonomy < ActiveRecord::Base
   }
+
+  def ancestor_ids
+    Taxonomy.find(taxonomy_id).ancestor_ids
```
why not using relation? like `taxonomy.ancestor_ids` or do we need clear object here?

ares commented on app/models/taxable_taxonomy.rb:
```diff
@@ -11,4 +12,16 @@ class TaxableTaxonomy < ActiveRecord::Base
+    Taxonomy.find(taxonomy_id).ancestor_ids
+  end
+
+  # in case the controller callback avoid_duplicate_taxable doesn't work
```
Could you please explain this a bit? I don't know why do we need this. This causes strange situations for me. Suppose we have locations A < B < C, I already assigned A and B now I edit a record and I'm trying to add C. Everything seems normal but this newly created association is deleted immediately and silently. On the other hand if I try to add A, B, C at the same time it works.

My feeling is that you try to disable user adding taxonomies where parents are already assigned. I don't think it's a good idea because you can get to that state by different order of assigning. I think it would be better if all children were assigned automatically so user would already see them assigned. However same problem would appear with unassigning them, therefore children taxonomies should be disabled in UI.

Since this could get tricky very soon, I'd prefer different UI if we want to support nested taxonomies. In worse case at least let the user know why some taxonomy can't be assigned.

isratrade replied:
@area, I added this method `ensure_uniqueness_on_parent` in case the controller method `avoid_duplicate_taxable` doesn't work. 
https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/taxonomies_controller.rb#L8

At first I didn't understand, but I now I see what you mean by disabling child taxonomies in UI.

isratrade commented:
re-based per @area's review.  I added default_scope for `order(:label)` and removed `reorder()`
Still to be done is disabling child taxonomies when parent is selected.

area commented:
While I'm happy to take credit for things I haven't done, I think @ares probably deserves it more!

isratrade commented:
@domcleal, any idea for this failure?

isratrade commented:
thanks @area for crediting @ares :)

domcleal commented:
@isratrade the build shows a failure during a clean db:migrate:

<pre>
SQLite3::SQLException: no such column: label: SELECT "taxonomies"."id" FROM "taxonomies"  ORDER BY label/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/sqlite3-1.3.8/lib/sqlite3/database.rb:91:in `initialize'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/sqlite3-1.3.8/lib/sqlite3/database.rb:91:in `new'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/sqlite3-1.3.8/lib/sqlite3/database.rb:91:in `prepare'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/activerecord-3.2.16/lib/active_record/connection_adapters/sqlite_adapter.rb:246:in `block in exec_query'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/activerecord-3.2.16/lib/active_record/connection_adapters/abstract_adapter.rb:280:in `block in log'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/activesupport-3.2.16/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/activerecord-3.2.16/lib/active_record/connection_adapters/abstract_adapter.rb:275:in `log'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/activerecord-3.2.16/lib/active_record/connection_adapters/sqlite_adapter.rb:242:in `exec_query'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/activerecord-3.2.16/lib/active_record/relation/calculations.rb:184:in `pluck'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-0/gems/activerecord-3.2.16/lib/active_record/querying.rb:12:in `pluck'
/var/lib/workspace/workspace/test_develop_pull_request@2/database/sqlite3/puppet/3.0/ruby/1.9.3/db/migrate/20130908100439_delete_orphaned_records.rb:23:in `up'
</pre>

I suspect that the migration here (db/migrate/20130908100439_delete_orphaned_records.rb) is failing because it uses the Taxonomy class, which assumes the label is present in the DB, but it's added by a later migration.

isratrade commented:
@domcleal, I fixed the migration.  Can you please review my changes to squash 3597109 from today.

isratrade commented:
@ares, I removed the method `ensure_uniqueness_on_parent` on taxable_taxonomy.rb.  After talking with @abenari, we feel it's best to allow duplicates rather than have a validation that causes unnecessary hassle for the user.  There is really no harm in the CLI or API adding extra rows in taxable_taxonomies even if they aren't necessary or "don't make sense".  
`with_taxonomy_scope` will work fine with or without the extra rows.  The only side effect (and it may be a possible one) is if a parent taxable_taxonomy row is deleted, then the child row still remains selected.  It is possible to also destroy the descendant taxable_taxonomies when destroying the parent row, but I agree with @abenari that this may annoy users in some cases.

isratrade commented:
@abenari, I added http://projects.theforeman.org/issues/4328 to improve UI experience for inherited locs/orgs.

isratrade commented:
@domcleal, is jenkin's stuck. I pushed this PR about 3 to 4 hours ago

domcleal commented:
@isratrade I don't know, there were some odd errors.  Try amending the commit and re-pushing.

isratrade commented:
re-committed.

isratrade commented:
This is a bug but shouldn't prevent the patch from entering.  The side-effect is not huge. 
http://projects.theforeman.org/issues/4333

isratrade commented:
@abenari, we can also implement the OptGroup feature in multi-select.js.  I added an issue http://projects.theforeman.org/issues/4334

ares commented:
ok, I think I somehow understand the logic behind inheritance (tomorrow I won't again :-) my only concern is that it does not seem intuitive. Otherwise :+1 from me (if you also store those taxonomies used so other scopes can reuse them). As I understood there are issues created for other my comments.

isratrade commented:
rebased after recent commits

domcleal commented on app/assets/stylesheets/application.scss:
```diff
@@ -177,6 +177,9 @@ select{padding: initial;}
 .delete{color: #b94a48 !important; cursor: auto;}
+.inherited{color: #00618a !important; cursor: auto;}
+.descendants{color: #00618a !important; cursor: auto;}
+.used_by_hosts{color: #000 !important; font-weight: bold !important; cursor: auto;}
```
Where are these used, I don't see it?

isratrade replied:
these css classes are used in 
https://github.com/isratrade/foreman/blob/inherit_loc/vendor/assets/javascripts/jquery.multi-select.js#L509-538

domcleal commented on app/controllers/concerns/foreman/controller/taxonomies_controller.rb:
```diff
@@ -87,6 +93,9 @@ def destroy
       process_error
     end
+  rescue Ancestry::AncestryException
+    flash[:error] = _('Cannot delete group %{current} because it has nested groups.') % { :current => @taxonomy.label }
```
"group" should be org/loc

domcleal commented on app/controllers/concerns/foreman/controller/taxonomies_controller.rb:
```diff
@@ -182,4 +191,16 @@ def count_nil_hosts
+    # The inner_select in #with_taxonomy_scope
+    # params = {:location => {:domain_ids => [1,2,3], :subnet_ids => [] ..}
+    params[taxonomy_single].merge!(@taxonomy.inherited_ids) do |k, submitted_ids, inherited_ids|
+      return submitted_ids unless submitted_ids.kind_of?(Array) && inherited_ids.kind_of?(Array)
```
I'm not sure a return is valid within a block like this

domcleal commented on app/controllers/concerns/foreman/controller/taxonomies_controller.rb:
```diff
@@ -182,4 +191,16 @@ def count_nil_hosts
 
+  def avoid_duplicate_taxable
+    # (-) minus operator to subtract elements of arrays. remove the inherited ids from params[:location] so new
+    # rows are not saved in taxable_taxonomies if the parent already has a row
```
I think this would be better in the model

domcleal commented on app/helpers/home_helper.rb:
```diff
@@ -32,11 +32,11 @@ def menu_item_tag item
 
   def org_switcher_title
     title = if Organization.current && Location.current
-      Organization.current.to_label + "@" + Location.current.to_label
```
A host group's `to_label` method returns the label if it exists, we should mirror that API

domcleal replied:
@isratrade?

isratrade replied:
@domcleal, I moved `to_label` to NestedAncestryCommon

domcleal commented on app/helpers/layout_helper.rb:
```diff
@@ -80,7 +80,7 @@ def multiple_checkboxes(f, attr, klass, associations, options = {}, html_options
 
   # add hidden field for options[:disabled]
-  def multiple_selects(f, attr, associations, selected_ids, options={}, html_options={})
+  def multiple_selects(f, attr, associations, selected_ids, options={}, html_options={}, method_name = :to_label )
```
If we redefine to_label, this wouldn't be necessary?

isratrade replied:
@domcleal, this is now fixed.

domcleal commented on app/models/concerns/taxonomix.rb:
```diff
@@ -62,4 +86,52 @@ def add_current_taxonomy?(taxonomy)
+
+  protected
+
+  def taxonmy_foreign_key_conditions
```
typo, taxonomy

domcleal commented on app/models/concerns/taxonomix.rb:
```diff
@@ -15,22 +15,46 @@ module Taxonomix
+      inner_ids ||= inner_ids_loc if inner_ids_loc
+      inner_ids ||= inner_ids_org if inner_ids_org
+      # In the case of users we want the taxonomy scope to get both the users of the taxonomy and admins.
+      inner_ids << admin_ids if inner_ids && self == User
```
Maybe extract this to a method on the class, which we can override on `User`?  e.g. additional_unscoped_taxonomy_ids?

isratrade replied:
@domcleal, I prefer to re-factor this after the merge.  Plus, there is already issue #4329 to optimize the SQL in this method

domcleal replied:
ok, I updated the issue

domcleal commented on app/models/concerns/taxonomix.rb:
```diff
@@ -15,22 +15,46 @@ module Taxonomix
-      taxonomy_ids = Array.wrap(taxonomy).map(&:id).join(',')
-      "SELECT taxable_id from taxable_taxonomies WHERE taxable_type = '#{self.name}' AND taxonomy_id in (#{taxonomy_ids}) "
+    # default inner_method includes parents (path_ids)
+    def with_taxonomy_scope_override(loc = nil, org = nil, inner_method = :path_ids)
```
Why use `:path_ids` here?  (Why is this different to `:subtree_ids` above?)  This means on the new host form that for "Israel" you wouldn't see domains for Tel Aviv or Rack 7.

isratrade replied:
`path_ids` goes up the tree.  `subtree_ids` does down.  They are the same as `ancestor_ids` and `descendant_ids` but they include loc/org that calls the method.
Notice on line #52 that that no matter which inner_method, it is joined with `ancestor_ids`

``` ruby
      taxonomy_ids = (taxonomy.send(inner_method) + taxonomy.ancestor_ids).uniq
```

This means that that override using `path_ids` isn't doing much since both `path_ids` and `ancestor_ids` and going in the same direction.

Correct, on new  host form if "Israel" is the current location, you wouldn't see domains for Tel Aviv or Rack 7.  This was the result of the disucssion with @ohadlevy.  If a user wants to add a host in Israel/Tel Aviv/Rack7, then the current needs to be Rack7 or have it as a parent.

isratrade replied:
if this is not the case, that I wrote above, then there is no need to override `with_taxonomy_scope`.  @ohadlevy ???

domcleal replied:
So perhaps what we're getting at is that there are three different behaviours of inherited taxonomies:
1. hosts in child taxonomies are visible when you're in the context of the parent
2. taxonomix resources in the parent, current or child taxonomies are visible from all ancestry contexts
3. when creating a new host, only taxonomix resources in the same or parent taxonomies are visible

Now to analyse them...
- (1) makes sense - it's intuitive, as you said below
- The parent bit of (3) makes sense, you assume that resources (especially domains) can filter down to all taxonomies below the one you assign it to.
- The child bit of (2) follows from (1), so that makes sense.  If I'm browsing domains in Israel, I expect to see tlv.redhat.com.
- The parent bit of (2) follows from (3), since you're able to create hosts with them, you should be able to see them too.

So the oddity is that although I can see resources in child taxonomies while in the context of a parent, I'm not able to use them on the host creation form unless I'm in within that child taxonomy.

GregSutcliffe replied:
Ok, after chatting with Dominic, I'm thinking like this:

(1) Makes perfect sense - if I'm in the Israel/TLV context, I expect to see the Hosts in Israel/TLV/Rack7 - so here, with the User as the subject, we're expecting to use `subtree_ids`
(2) Dominic's analysis works for me, so we're saying that for non-host taxonomix objects, we're  doing something like `subtree_ids + ancestor_ids`
(3) This is the slightly tricky one - ensuring that we render the Host#new form with the correct options. My feeling is that we either:

a) restrict Hosts to being created in a leaf-node Taxonomy, or
b) use the same `subtree_ids + ancestor_ids` code to look both ways along the tree and allow Hosts to be created with resources for either direction.

I'm leaning towards (b) for greater flexibility. It may be that a user wishes to create a host in Israel/TLV that uses an Install Media in Israel/TLV/Rack7.

isratrade replied:
@GregSutcliffe, For clarification, @domcleal wrote "I'm not able to use them on the host creation form unless I'm in within that child taxonomy".  The key point is that "within that child taxonomy" does not mean that a user needs to go to the org/loc picker (that defines Location.current, Organization.current), but rather the taxonomies selected for the new host on the host form

isratrade replied:
@GregSutcliffe, @domcleal, I spoke with @ohadlevy  and @abenari  and they prefer to keep the logic as is that when creating a new host and after selecting loc and org, the drop-down boxes only show what's available in loc/org or inherited from parents

domcleal replied:
Should we do solution (a) from Greg's comment then to make this less confusing?

isratrade replied:
@domcleal. I don't think it is practical to restrict hosts to being created in a leaf-node taxonomy.  This means that adding a child taxonomy will prevent future host creation in the parent, as far as I understand what @GregSutcliffe means by option (a)

domcleal replied:
Yeah, that's a very good point.  So why couldn't we also show resources from child taxonomies?  I'd be interested in the reasoning, as it seems disjointed at the moment.

ohadlevy replied:
Given the following example:

Location(s)
 Europe / Czech / Brno
 Europe / Czech / Prague

Brno has Subnet A
Prague has Subnet B
Czech has domain cz.corp.com

Location.current = Any Context / Eurpoe (doesnt matter)

If you would select location (in the host new form) as Czech, you wont be
able to save the host, as it has no subnets.

Would you prefer to see both Subnet A and Subnet B instead? (sounds less
practical to me).

domcleal replied:
@ohadlevy yes, that's what I'd expect.  Especially as if I click Infrastructure > Subnets, I see both Subnet A and B.

GregSutcliffe replied:
Agreed, I'd expect to be able to select the subnets from the children. Without that, you're _effectively_ creating the leaf-node restriction I talked about (for this example anyway, not generally true) - since the valid subnets are only on the leaf nodes, hosts can only be created there.

ohadlevy replied:
I tend to disagree, as you should select the specific target you are
deploying to first. otherwise, whats the point in allocating resources to
child taxonomy ?

in a large env (e.g. where location is every rack) this would be
unmanageable at best.

domcleal replied:
Okay, we had a meeting to discuss these issues and came to the following conclusions:
1. keep this behaviour as currently implemented
2. in the web UI, investigate ways to improve the display of resource lists that currently show resources both up and down the tree, e.g.
   1. filters that allow the user to select "upwards" or "downwards" views
   2. allow multiple current taxonomies, so users can select an entire tree of current taxonomies
   3. encourage users to stay in "any context" and rely on search techniques more
   4. add the taxonomies for each resource to the list
3. in the CLI, investigate ways to show available resources for host creation, since "hammer domain list" will show all domains, not necessarily those relevant for host creation in a given context

We generally felt that the host index was intuitive, looking "down" only and that looking both up and down made sense for resources, so you could see things that had filtered down to you, and that were in your context's children.

For host creation, users should be encouraged to create hosts in leaf taxonomies, or to assign resources higher up the taxonomy so they're available.

domcleal replied:
Filed http://projects.theforeman.org/issues/4376 for the UI and http://projects.theforeman.org/issues/4377 for the CLI.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -55,8 +55,8 @@ class Jail < ::Safemode::Jail
       conditions = {}
-      conditions[:organization_id] = Array.wrap(org).map(&:id) if org
-      conditions[:location_id]     = Array.wrap(loc).map(&:id) if loc
+      conditions[:organization_id] = org.subtree_ids if org
```
[done] path_ids?  So if your current loc is Tel Aviv, you also see Israel hosts.

isratrade replied:
This was the change from my first demo, which had `path_ids` and was not intuitive.  Israel/Tel Aviv/Rack7 showed 35 hosts, but Israel/Tel Aviv only showed 26 hosts.  Now, it's opposite. Israel/Tel Aviv shows all 35 hosts and Israel/Tel Aviv/Rack7 shows 9 hosts that only are assigned to Rack 7.   Hosts are unique since they are the only object that belongs_to one location or organization.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -346,7 +346,7 @@ def host_inherited_params include_source = false
     if SETTINGS[:locations_enabled] && location
       # read location parameters
-      location.parameters.each {|p| hp.update Hash[p.name => include_source ? {:value => p.value, :source => N_('location').to_sym} : p.value] }
+      location.selected_or_inherited_parameters.each {|p| hp.update Hash[p.name => include_source ? {:value => p.value, :source => N_('location').to_sym} : p.value] }
```
[done] organization hasn't been updated above to do the same

domcleal commented on app/models/taxonomies/location.rb:
```diff
@@ -25,4 +25,16 @@ def dup
+  end
+
+  def selected_or_inherited_parameters
+    LocationParameter.where(:reference_id => path_ids)
```
This isn't as intelligent as `Hostgroup#parameters`.  The hg version will iterate over the ancestry in order and override parameters as it gets deeper, while this implementation won't - we should unify them.

isratrade replied:
@domcleal, to re-factor this efficiently, I either need to cherry-pick the changes to NestedAncestryCommon in PR ##1155 or wait until it's merged.  Or, I can temporarily duplicate code from hostgroup.rb here in location.rb if this PR is going to be merged first.

domcleal replied:
I'll do 1155 first then, thanks.

domcleal commented on app/models/smart_proxy.rb:
```diff
@@ -74,6 +74,14 @@ def refresh
+    feature_names = self.features.pluck(:name)
+    conditions                      = {}
+    conditions[:puppet_proxy_id]    = id if feature_names.include?('Puppet')
+    conditions[:puppet_ca_proxy_id] = id if feature_names.include?('Puppet CA')
```
I think this will generate an `AND`, so you'd only pull back hosts that have both a puppet proxy and a puppet CA proxy set to this resource, when in fact you probably want OR.

domcleal commented on app/models/taxonomies/location.rb:
```diff
@@ -25,4 +25,16 @@ def dup
     new
   end
 
+  def lookup_value_match
```
Unnecessary?  Isn't this only needed if we had puppet classes associated to the location?

If it is needed, it could be in Taxonomy.

isratrade replied:
It's used in host_common.rb #22
https://github.com/isratrade/foreman/blob/inherit_loc/app/models/concerns/host_common.rb#L22

``` ruby
    has_many :lookup_values, :finder_sql => Proc.new { LookupValue.where('lookup_values.match' => lookup_value_match).to_sql }, :dependent => :destroy
```

If it's in taxonomy.rb, then the matcher needs to determine the sti type. I think this is easier to understand.

``` ruby
    "location=#{label}"
```

domcleal commented on app/models/taxonomy.rb:
```diff
@@ -21,12 +23,15 @@ class Taxonomy < ActiveRecord::Base
   has_many :subnets, :through => :taxable_taxonomies, :source => :taxable, :source_type => 'Subnet'
 
-  scoped_search :on => :name, :complete_value => true
+  scoped_search :on => :label, :complete_value => true
```
Host groups can be searched by name and label, we should keep both?  This doesn't seem to work on the org/loc index page btw.

domcleal commented on app/views/common_parameters/_inherited_parameter.html.erb:
```diff
@@ -0,0 +1,17 @@
+    </td>
+    <td class="col-md-1">
+    <span class="help-block">
+      <%= "Inherited" %>
```
i18n extraction

domcleal replied:
Optional, the new host form has an Override button which lets you override the inherited value by creating a new parameter with the same name, it'd be a nice-to-have here.  (Related, this should be possible on host groups too, but the form doesn't show them, [#4348](http://projects.theforeman.org/issues/4348)).

domcleal commented on app/views/common_parameters/_inherited_parameter.html.erb:
```diff
@@ -0,0 +1,17 @@
+      <%= f.text_field(:name, :value => param.name, :class => "form-control", :disabled => true) %>
+    </td>
+    <td class="col-md-7">
+      <%= f.text_area(:value, :value => param.name, :class => "form-control", :rows => line_count(f, :value), :disabled => true) %>
```
Text field value is the parameter name, not the parameter value

domcleal commented on app/views/domains/_form.html.erb:
```diff
@@ -1,4 +1,4 @@
-<%= form_for @domain, :url => (@domain.new_record? ? domains_path : domain_path(:id => @domain.id)) do |f| %>
+<%= form_for @domain do |f| %>
```
This undoes 696420ff?

isratrade replied:
@domcleal, oh yes, I remember this fix now.  I did it :)

domcleal commented:
By the way, you have some integration test failures.

One more difference I noticed between this and host groups is that the parent host group is shown in a select box when you edit it.  Could we add this to the taxonomy edit page?  I think it'd be safe to make it editable even, but if there's an issue, then it could be left disabled.

isratrade commented on app/models/taxonomies/location.rb:
```diff
@@ -16,13 +16,32 @@ class Location < Taxonomy
+        where(conditions)
       }
 
+  scoped_search :on => :label, :complete_value => :true, :default_order => true
```
this was added to location.rb and organization.rb even though it appears in the NestedAncestryCommon mix. Otherwise, the autocompleter won't find "label = "

isratrade commented:
I assume the failure on Ruby 1.8.7 sqlite3 only is from jenkins.

domcleal commented on app/controllers/concerns/foreman/controller/taxonomies_controller.rb:
```diff
@@ -87,6 +95,9 @@ def destroy
     else
       process_error
     end
+  rescue Ancestry::AncestryException
```
[done] Can you add this to the taxonomies API controller concern too please?

isratrade replied:
yes, added to api controller

domcleal commented on app/models/taxonomies/location.rb:
```diff
@@ -16,13 +16,32 @@ class Location < Taxonomy
+  end
+
+  def sti_name
+    N_("location")
```
This should be `_()` else it won't get translated (N_ returns the original string and you're not translating it in the controller).

isratrade commented:
re-factored and with all fixes to comments.

isratrade commented on app/controllers/concerns/foreman/controller/taxonomies_controller.rb:
```diff
@@ -70,8 +75,7 @@ def step2
 
   def update
     result = Taxonomy.no_taxonomy_scope do
-      (params[taxonomy_single.to_sym][:ignore_types] -= ["0"]) if params[taxonomy_single.to_sym][:ignore_types]
```
this is now handle in the `before_validation` in the model

isratrade commented on app/services/tax_host.rb:
```diff
@@ -117,12 +126,16 @@ def check_for_orphans
     !found_orphan
   end
 
+  def non_inherited_ids(v1 = self.selected_ids, v2 = self.inherited_ids)
```
this method `non_inherited_ids` is not used now that I we don't send all disabled selected_ids, but I prefer to keep it

isratrade commented on app/models/hostgroup.rb:
```diff
@@ -68,11 +65,6 @@ def all_puppetclasses
     classes
   end
 
-  def to_label
```
this was moved to NestedAncestryCommon mixin

isratrade commented on app/models/hostgroup.rb:
```diff
@@ -10,8 +10,7 @@ class Hostgroup < ActiveRecord::Base
   has_many :puppetclasses, :through => :hostgroup_classes
   has_many :user_hostgroups, :dependent => :destroy
   has_many :users, :through => :user_hostgroups
-  validates :name, :uniqueness => {:scope => :ancestry, :case_sensitive => false },
```
this was moved to NestedAncestryCommon mixin

isratrade commented on app/models/taxonomies/location.rb:
```diff
@@ -16,13 +16,32 @@ class Location < Taxonomy
         else
           conditions = sanitize_sql_for_conditions([" (taxonomies.id in (?))", user.location_ids])
         end
-        where(conditions).reorder('type, name')
```
@abenari, do you recall why you put `.reorder('type, name')` here?  If yes, then I can put it back.

isratrade commented:
@domcleal, if I use the same code for inherited parameters in locs/orgs as in hostgroups, then it doesn't show inherited values.  Should this bug http://projects.theforeman.org/issues/4348 need to prevent the merger of this PR?

domcleal commented:
@isratrade no, let's fix that later, it's only a UI issue

isratrade commented:
@ohadlevy, removed unused view partial and re-commited

Commit:
fixes #3912 - add inheritance for locations / organizations

isratrade commented:
rebased after multi-select commit

domcleal commented:
Thanks @isratrade!  Merged as 1fa008a for Foreman 1.5.  Nice work :)
